src/net_processing.cpp

```diff
@@ -1605,7 +1622,13 @@ void static ProcessGetData(CNode* pfrom, const CChainParams& chainparams, CConnm
                     push = true;
                 }
             }
-            if (!push) {
+
+            if (push) {
```

Comment:
Imagine in future we will not care about privacy here (e.g., use Tor or whatnot), and allow `push` for unsolicited GETDATA.
Then, if the condition for rebroadcast is still `if (push)` it would be trivial to censor a transaction by sending unsolicited GETDATA. It means that with this code, effectiveness of rebroadcast depend on this privacy behaviour we have.
I'm afraid that this non-obvious relation might screw us up in future. I wish we documented this more, maybe in the comment right above.
I'm fine with keeping it this way too, just want to raise some awareness here.

Reply:
Sorry, I don't follow. How would a node send GETDATA without knowing the txid? Also, how would it censor (prevent you other connections from sending a GETDATA) by just sending a GETDATA?

Reply:
I'm not fully following either, but a couple of points to highlight: 
- lets distinguish *unbroadcast* from *rebroadcast*. unbroadcast is extra attempts at successful initial broadcast. rebroadcast is repeated broadcasts (from the POV of the node, which cannot know full picture). 
- in this implementation, if victim sends INV to only one peer, peer is adversarial, replies back with GETDATA but doesn't propagate it out to the network -> victim removes from unbroadcast set & doesn't retry until ~1 day later when `ResendWalletTransactions` is triggered. so, prevents txn from making it out to mempool for a day.
- @naumenkogs do you think unsolicited GETDATA messages would change this?

Reply:
Imagine a malicious node just found out that some wallet has to do broadcast soon. Imagine a lightning node submits a tx, but it is offline, and will be online in 30 minutes to relay it from Unbroadcast.
Now an attacker connect to the victim once they got online, and send GETDATA for that transaction (before a victim broadcasts from unbroadcast).

Right now, it will be ignored, because it's unsolicited (for privacy reasons!).
If we discard those privacy reasons, and now forget about this *inherent unbroadcast behaviour*, and allow unsolicited GETDATA, an attacker will be able to prevent us from broadcasting a tx.

And yeah, one day for lightning is pretty bad, you will simply loose money in most of the implementations today.

Reply:
Yes that's a good point, in LN there is class of transaction where txid is known from remote party due to the fact they counter-sign them (HTLC-timeout) but honest party have to broadcast with time-constrain to avoid any risk of money-loss.

Right now most of implementations are pretty aggressive at triggering rebroadcast at every block, i.e calling again `sendrawtransaction` which isn't great for privacy and this may change in the future. That said, often offline nodes, like mobile, won't use core as a backend and broadcast directly their transaction on the p2p layer or submit to a third-party server. Routing nodes assumed to run core, should rebroadcast at every block so it would override some attacker messing with their unbroadcast logic.

I think adopting wtxid would solve this issue because counterparty won't know about witness, at least for LN. But I agree, we should document this to keep track of this (add a comment commit on top or as a follow-up)

Reply:
A simpler fix would be to add a one-line check that `pfrom` is outbound, no?

Reply:
I think that's okay, it won't degrade the reliability improvement of aimed by this PR because we already announce sooner to outbound. But the broader we should document better these maybe-cases of time-sensitive tx propagation obstruction (both at p2p/mempool layers) like last thread on ml shows.

Reply:
ok, updating to check `pfrom` is from outbound sounds good to me. increases delivery guarantee for the general case too. I added this to the todo list on follow ups pr https://github.com/bitcoin/bitcoin/pull/18807#issue-410374742. code is just 1 line change, but I want to make sure to update other relevant comments to clarify its removed via _outbound_ getdata & hopefully add a test case. 

I'm going to resolve this comment here. thanks!

Reply:
on further thought, I don't think it makes sense to only remove unbroadcast txns on receiving `GETDATA` from outbound peers. 

this relies on the expectation that outbound peers get the txn before inbound peers (otherwise they might relay), which isn't necessarily true in edge cases. if we were to implement, I'd want to add logic to remove from unbroadcast set if we receive an `inv` from a peer. However, doing all this seems unnecessary right now, especially considering the greater direction of this project (#16698).